### ISSUE TITLE
Procedures with const parameters

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#297]
Currently, we for some built-in assembly instructions we allow specifying immediate values (e.g., `add.b`). This is convenient because in many cases, knowing a value which the instruction needs to work with at compile time allows for more efficient code generation.

With procedures, this is currently not possible. So, we can't say something like `u64::checked_add.b`. In some cases, this leads to much less efficient code (e.g., for bitwise shifts and rotations). We could address this by allowing procedures to receive constant parameters.

To support this, we need to change a couple of things:

First, we need to make sure constant parameters can be specified at procedure declaration time. For example, something like this could work:
```
proc.foo<a> # single parameter
proc.foo<a,b> # multiple parameters (no whitespace between parameters)
proc.foo<a>.2 # procedure with const parameter and locals
```

Next, we need to specify how these parameters could be used with other instructions. I'm thinking something as simple as this could work:
```
proc.foo<a,b>
  add.a
  add.b
end
```

Lastly, we need to specify how to invoke these procedures. Here again, I think it could be quite simpler:
```
exec.foo.1 # if foo has 1 const parameter
exec.foo.1.2 # if foo has multiple const parameters
```

**Comment by skaller** [comment]
So roughly, procedures are now also macros? This certainly allows for better code. However, I don't see exactly how the same code would work if you didn't provide some of the constant parameters and wanted the values to be on the stack instead.

```
proc.foo<a,b>
  add.a
  add.b
end
```

Fine, but what if I want 
```
push.42
exec.foo.97
```
does the `.b` just fall off the `add.b`? Will that always get the right value on the stack? No!

If you do have 'constant parameters` they will always have to be constants. However, the values used in the program can now be any constant expression:
```
proc.foo<a,b>
  add.<a+b> # add constants
end
```
There's a danger  I think, that improving the assembler could make writing a front end to it harder.

There is a language which is a total master of this kind of idea, where constants known at compile time and values only known at run time can be arbitrarily mixed and use the same syntax for compile time and run time calculations, and that's modern C++. Just throw `constexpr` around everywhere and your ackermann's function implementation can happily run at compile time and the result can be used as an array length.

The best way to obtain the kind of effect desired here is to design a high level language and write an optimising compiler. It's probably going to inline your procedure anyhow and use constants where it has an actual performance impact.

**Comment by bitwalker** [comment]
@bobbinth I don't think we actually want this, or at least, I don't think it's actually worth the complexity it introduces. Ultimately, this sort of thing should really be the purview of compilers, rather than the assembler.

If we really felt like something along these lines was valuable, then some form of light macro system would be better, where you can actually do something useful based on the constant inputs - but again, I'm just not convinced that's really worth the cost at the assembly level.

I'd recommend we close this issue as not planned.

**Comment by bobbinth** [comment]
I think optimizations like this are going to be equally complex in the compiler, but agreed that the complexity is probably not worth it. At least, this is not something we'll do in the foreseeable future - so, closing this for now.
